### PR TITLE
Reject unsupported `f_divergence_type` / loss type combinations in DPO

### LINKED ISSUE #6441: DPOTrainer apo_down loss inconsistently applies f_divergence_type across its two terms
### Reproduction

`DPOTrainer`'s `apo_down` loss (`trl/trainer/dpo_trainer.py`, lines 1509-1515) treats its two terms inconsistently with respect to `f_divergence_type`:

```python
elif loss_type == "apo_down":
    # Eqn (8) of the APO paper (https://huggingface.co/papers/2408.06266)
    losses_chosen = torch.sigmoid(self.beta * chosen_logratios)      # always the raw log-ratio
    losses_rejected = 1 - torch.sigmoid(self.beta * delta_score)     # f-divergence-transformed
    per_sequence_loss = losses_chosen + losses_rejected
```

`chosen_logratios`/`rejected_logratios` (lines 1384-1385) are the plain `log(π_θ/π_ref)`. `delta_score` (line 1417) is `chosen_scores - rejected_scores`, where `chosen_scores`/`rejected_scores` (lines 1387-1415) are only equal to the raw logratios when `f_divergence_type="reverse_kl"` (the default); for `forward_kl`, `js_divergence`, or `alpha_divergence` they're a nonlinear transform of the logratios.

Checked this against the actual paper: APO (arXiv:2408.06266), Eq. 8 (extracted directly from the paper's own LaTeX source via `ar5iv.labs.arxiv.org/html/2408.06266`):

```
L_down^APO(x, yw, yl; θ) = σ(r_θ(x,yw)) − σ(r_θ(x,yw) − r_θ(x,yl))
```

where `r_θ(x,y) = β·log(π_θ(y|x)/π_ref(y|x))` (the paper's Eq. 5) is always the plain reverse-KL log-ratio, never any other f-divergence.

`apo_zero`, two branches above (lines 1501-1507), correctly keeps both terms on the raw log-ratio regardless of `f_divergence_type`:
```python
losses_chosen = 1 - torch.sigmoid(self.beta * chosen_logratios)
losses_rejected = torch.sigmoid(self.beta * rejected_logratios)
```
`apo_down` should be symmetric with `apo_zero` here, but its second term uses `delta_score` instead, almost certainly because `delta_score` happens to equal `chosen_logratios - rejected_logratios` in the default `reverse_kl` case, which hides that it silently changes meaning once `f_divergence_type` is set to anything else. There's no validation that rejects this combination, `loss_type=["apo_down"]` with a non-default `f_divergence_type` is a legal, silent configuration.

```python
import torch
import torch.nn.functional as F

beta = 0.1
chosen_logratios = torch.tensor([0.5, -0.3, 1.2])
rejected_logratios = torch.tensor([-0.2, 0.4, -0.6])

def scores_for(f_divergence_type, chosen_logratios, rejected_logratios):
    if f_divergence_type == "reverse_kl":
        return chosen_logratios, rejected_logratios
    elif f_divergence_type == "js_divergence":
        return F.logsigmoid(chosen_logratios), F.logsigmoid(rejected_logratios)

def apo_zero_loss(chosen_logratios, rejected_logratios):
    losses_chosen = 1 - torch.sigmoid(beta * chosen_logratios)
    losses_rejected = torch.sigmoid(beta * rejected_logratios)
    return losses_chosen + losses_rejected

def apo_down_loss_as_coded(chosen_logratios, delta_score):
    losses_chosen = torch.sigmoid(beta * chosen_logratios)
    losses_rejected = 1 - torch.sigmoid(beta * delta_score)
    return losses_chosen + losses_rejected

# apo_zero is internally consistent across f_divergence_type (never touches chosen_scores/rejected_scores)
apo_zero_reverse_kl = apo_zero_loss(chosen_logratios, rejected_logratios)
print("apo_zero identical regardless of f_divergence_type:", True)  # by construction

# apo_down is NOT
chosen_scores, rejected_scores = scores_for("js_divergence", chosen_logratios, rejected_logratios)
delta_score = chosen_scores - rejected_scores
code_result = apo_down_loss_as_coded(chosen_logratios, delta_score)

def apo_down_fully_generalized(chosen_scores, rejected_scores):
    delta = chosen_scores - rejected_scores
    losses_chosen = torch.sigmoid(beta * chosen_scores)
    losses_rejected = 1 - torch.sigmoid(beta * delta)
    return losses_chosen + losses_rejected

fully_generalized = apo_down_fully_generalized(chosen_scores, rejected_scores)
print("code's apo_down under js_divergence:      ", code_result)
print("fully f-divergence-generalized apo_down:   ", fully_generalized)
print("max abs diff:", (code_result - fully_generalized).abs().max().item())
```

Output:
```
apo_zero identical regardless of f_divergence_type: True
code's apo_down under js_divergence:       tensor([1.0044, 1.0010, 1.0106])
fully f-divergence-generalized apo_down:    tensor([0.9800, 0.9872, 0.9741])
max abs diff: 0.03654569387435913
```

### System Info

TRL commit `65eb0df` (current `main`), reproduced with pure `torch` tensors, no GPU or model download needed since this is testing the loss math directly.

### Checklist

- [x] I have checked that my issue isn't already filed
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible

### Suggested fix

Make `apo_down`'s second term consistent with `apo_zero`'s pattern, either always compute it from `rejected_logratios` regardless of `f_divergence_type` (matching the paper's Eq. 8, which never defines `r_θ` via any f-divergence other than reverse-KL), or, if a fully-f-divergence-generalized version of APO-down is intended, also change the first term to use `chosen_scores` instead of `chosen_logratios` so both terms are consistently transformed. Right now it's neither, one term is raw, the other is transformed, which matches neither reading.

### PR DESCRIPTION
**Pull request by qgallouedec**
Fixes #6441. Closes #6475, which patched `apo_down` alone.

The f-DPO reparameterization leaves a `β·log Z(x)` term that only cancels in the chosen-rejected reward difference, so `f_divergence_type` is only meaningful for difference-based losses. Today the others either silently ignore it or, worse, are half-wired to it.

- Raise at init when `f_divergence_type != "reverse_kl"` and any loss type is outside `{sigmoid, sigmoid_norm, hinge, ipo, exo_pair, robust, discopop, sft}`.
- Use raw log-ratios in `nca_pair`, `bco_pair`, `sppo_hard` and `apo_down`, so the loss chain reads the same boundary as the validator.

## History

Before #3906, `f_divergence_type` reached only the pairwise `logits`; losses that recomputed their own quantities never saw it. That refactor introduced `chosen_scores` / `rejected_scores` / `delta_score` and widened the reach; deliberately for some losses, accidentally for others.

| loss | before #3906 | after #3906 | this PR |
|---|---|---|---|
| `sigmoid`, `hinge`, `ipo`, `exo_pair`, `robust` | f-div | f-div | unchanged |
| `discopop` | raw (recomputed `logits` locally) | f-div | unchanged, keeps f-div |
| `sigmoid_norm` | did not exist | f-div | unchanged |
| `nca_pair`, `bco_pair`, `sppo_hard` | raw | f-div | back to raw |
| `apo_down` | raw | **mixed: raw chosen, f-div rejected** | back to raw |
| `apo_zero`, `aot`, `aot_unpaired` | raw | raw | unchanged |
| `sft` | n/a | n/a | unchanged |

All edits are identity under the default `reverse_kl`! Breaking only for `nca_pair` / `bco_pair` / `sppo_hard` / `apo_*` / `aot*` with a non-default divergence.

---

> [!NOTE]
> **Medium Risk**
> Changes core DPO loss computation and can alter training when using non-default f_divergence with nca_pair, bco_pair, sppo_hard, or apo_* losses; default reverse_kl paths are unchanged.
> 
> **Overview**
> Tightens **f-DPO** (`f_divergence_type`) support in `DPOTrainer` so unsupported combinations fail at init instead of being ignored or half-applied.
> 
> **Validation:** Non-default `f_divergence_type` is allowed only with difference-based losses (`sigmoid`, `sigmoid_norm`, `hinge`, `ipo`, `exo_pair`, `robust`, `discopop`, `sft`). Other losses (`apo_*`, `aot*`, `nca_pair`, `sppo_hard`, etc.) must use `reverse_kl`. **`use_liger_kernel=True`** with a non-default divergence now raises, since Liger only implements reverse-KL.
> 
> **Loss math:** `forward_kl`, `js_divergence`, and `alpha_divergence` scoring no longer drop terms that belong in the f′ transform. **`nca_pair`**, **`bco_pair`**, **`sppo_hard`**, and **`apo_down`** again use raw log-ratios (fixing accidental f-div wiring and the **`apo_down`** bug where the rejected term used f-div scores). Default `reverse_kl` behavior is unchanged for standard DPO setups.
> 
> **Docs/tests:** `DPOConfig` help text documents the loss whitelist; new tests cover init failures for bad loss/divergence pairs and Liger + non-default divergence.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 426d0bea36dfdfafc9847b2a3d0ec7cf16e95f54. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -768,6 +768,15 @@ def __init__(
         self.ld_alpha = args.ld_alpha
         self.f_divergence_type = args.f_divergence_type
         self.f_alpha_divergence_coef = args.f_alpha_divergence_coef
+        # The f-DPO reparameterization leaves a β·log Z(x) term that only cancels in the chosen-rejected reward
+        # difference, so the other losses have no valid f-divergence generalization.
+        f_divergence_loss_types = {"sigmoid", "sigmoid_norm", "hinge", "ipo", "exo_pair", "robust", "discopop", "sft"}
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:774]: fair, but deferring it, because I need to double check, and it's not urgent/critical

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1409,13 +1419,13 @@ def _compute_loss(self, model, inputs, return_outputs):
...
+            # f'(t) = 1 - 1/t
+            chosen_scores = 1 - torch.exp(-chosen_logratios)
+            rejected_scores = 1 - torch.exp(-rejected_logratios)
         elif self.f_divergence_type == "js_divergence":
-            # f'(t) = log(2t/(t+1)) -> drop log 2
-            chosen_scores = F.logsigmoid(chosen_logratios)
-            rejected_scores = F.logsigmoid(rejected_logratios)
+            # f'(t) = log(2t/(t+1)) = log 2 + logsigmoid(log t)
+            chosen_scores = math.log(2) + F.logsigmoid(chosen_logratios)
+            rejected_scores = math.log(2) + F.logsigmoid(rejected_logratios)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1434]: The additive constants of f' are kept: they cancel in the chosen-rejected difference, but `ipo` and `sigmoid_norm` divide each score by its own completion length before subtracting, so dropping them would leave a length-dependent bias.

**Review by albertvillanova — APPROVED** [review]
